// @adonisjs/core/build/ignitor-CEEHASdH.js
import { t as Ignitor } from "./main-zFY2qkGC.js";
import { t as logger_exports } from "./logger-Cj7AOEsL.js";
import { t as main_exports } from "./main-BDd3QiJF.js";
import { n as drivers, t as defineConfig } from "./define_config--5euEdi8.js";
import { n as defineConfig$1, r as drivers$1 } from "./main-CG-utHjr.js";
import { t as main_exports$1 } from "./main-DpE5rTyk.js";
//#region factories/core/ignitor.ts
/**
* Factory for creating and configuring AdonisJS Ignitor instances.
* This factory provides a fluent API to set up applications with core providers,
* configurations, and preload actions for testing and development scenarios.
*
* @example
* ```ts
* const ignitor = new IgnitorFactory()
*   .withCoreProviders()
*   .withCoreConfig()
*   .preload((app) => {
*     // Custom initialization logic
*   })
*   .create(new URL('../', import.meta.url))
*
* const app = ignitor.createApp('web')
* await app.boot()
* ```
*/
var IgnitorFactory = class {
	#preloadActions = [];
	#parameters = {};
	/**
	* A flag to know if we should load the core providers
	*/
	#loadCoreProviders = false;
	/**
	* Define preload actions to run during application initialization.
	* These actions are executed after the application is booted.
	*
	* @param action - Function to execute during preload phase
	*
	* @example
	* ```ts
	* factory.preload((app) => {
	*   // Register custom bindings
	*   app.container.bind('customService', () => new CustomService())
	* })
	* ```
	*/
	preload(action) {
		this.#preloadActions.push(action);
		return this;
	}
	/**
	* Merge core providers with user defined providers
	*/
	#mergeCoreProviders(providers) {
		return [
			() => import("@adonisjs/core/providers/app_provider"),
			() => import("@adonisjs/core/providers/hash_provider"),
			() => import("@adonisjs/core/providers/repl_provider")
		].concat(providers || []);
	}
	/**
	* Merge custom factory parameters with existing ones.
	* This allows you to customize RC file contents and application configuration.
	*
	* @param params - Parameters to merge
	* @param params.config - Application configuration to merge
	* @param params.rcFileContents - RC file contents to merge
	*
	* @example
	* ```ts
	* factory.merge({
	*   config: {
	*     database: { connection: 'mysql' }
	*   },
	*   rcFileContents: {
	*     commands: ['./commands/CustomCommand']
	*   }
	* })
	* ```
	*/
	merge(params) {
		if (params.config) this.#parameters.config = Object.assign(this.#parameters.config || {}, params.config);
		if (params.rcFileContents) this.#parameters.rcFileContents = Object.assign(this.#parameters.rcFileContents || {}, params.rcFileContents);
		return this;
	}
	/**
	* Include core AdonisJS providers when booting the application.
	* This adds essential providers like app, hash, and REPL providers.
	*
	* @example
	* ```ts
	* const ignitor = new IgnitorFactory()
	*   .withCoreProviders()
	*   .create(appRoot)
	* ```
	*/
	withCoreProviders() {
		this.#loadCoreProviders = true;
		return this;
	}
	/**
	* Merge default configuration for core AdonisJS features.
	* This includes configurations for HTTP, hash, logger, and bodyparser.
	* A shallow merge is performed with existing config.
	*
	* @example
	* ```ts
	* const ignitor = new IgnitorFactory()
	*   .withCoreConfig()
	*   .create(appRoot)
	* ```
	*/
	withCoreConfig() {
		this.merge({ config: {
			app: {
				appUrl: "http://localhost:3333",
				http: (0, main_exports.defineConfig)({})
			},
			validator: {},
			encryption: defineConfig({
				default: "gcm",
				list: { gcm: drivers.aes256gcm({
					id: "gcm",
					keys: ["averylongrandomsecretkey"]
				}) }
			}),
			bodyparser: (0, main_exports$1.defineConfig)({}),
			hash: defineConfig$1({
				default: "scrypt",
				list: { scrypt: drivers$1.scrypt({}) }
			}),
			logger: (0, logger_exports.defineConfig)({
				default: "app",
				loggers: { app: {} }
			})
		} });
		return this;
	}
	/**
	* Create a configured Ignitor instance with all specified parameters.
	*
	* @param appRoot - Application root directory URL
	* @param options - Optional Ignitor configuration options
	*
	* @example
	* ```ts
	* const ignitor = new IgnitorFactory()
	*   .withCoreConfig()
	*   .withCoreProviders()
	*   .create(new URL('../', import.meta.url))
	* ```
	*/
	create(appRoot, options) {
		return new Ignitor(appRoot, options).tap((app) => {
			app.booted(async () => {
				for (let action of this.#preloadActions) await action(app);
			});
			if (this.#loadCoreProviders) {
				this.#parameters.rcFileContents = this.#parameters.rcFileContents || {};
				this.#parameters.rcFileContents.providers = this.#mergeCoreProviders(this.#parameters.rcFileContents.providers);
			}
			this.#parameters.rcFileContents && app.rcContents(this.#parameters.rcFileContents);
			this.#parameters.config && app.useConfig(this.#parameters.config);
		});
	}
};
//#endregion
export { IgnitorFactory as t };
