// @adonisjs/core/build/main-BDd3QiJF.js
import { n as __reExport, t as __exportAll } from "./chunk-ZRW5x4aP.js";
import "@adonisjs/bodyparser";
//#region modules/http/request_validator.ts
/**
* Request validator for validating HTTP request data using VineJS validators.
* This class provides a convenient way to validate request body, files, cookies,
* headers, and URL parameters in AdonisJS applications.
*
* @example
* ```ts
* // Inside a controller method
* const data = await request.validateUsing(createUserValidator, {
*   messagesProvider: customMessages
* })
* ```
*/
var RequestValidator = class RequestValidator {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	/**
	* The error reporter method returns the error reporter
	* to use for reporting errors.
	*
	* You can use this function to pick a different error reporter
	* for each HTTP request
	*/
	static errorReporter;
	/**
	* The messages provider method returns the messages provider to use
	* finding custom error messages
	*
	* You can use this function to pick a different messages provider for
	* each HTTP request
	*/
	static messagesProvider;
	#requestData() {
		return {
			...this.#ctx.request.all(),
			params: this.#ctx.request.params(),
			headers: this.#ctx.request.headers(),
			cookies: this.#ctx.request.cookiesList()
		};
	}
	#processValidatorOptions(options) {
		const validatorOptions = options || {};
		/**
		* Assign request specific error reporter
		*/
		if (RequestValidator.errorReporter && !validatorOptions.errorReporter) {
			const errorReporter = RequestValidator.errorReporter(this.#ctx);
			validatorOptions.errorReporter = () => errorReporter;
		}
		/**
		* Assign request specific messages provider
		*/
		if (RequestValidator.messagesProvider && !validatorOptions.messagesProvider) validatorOptions.messagesProvider = RequestValidator.messagesProvider(this.#ctx);
		return validatorOptions;
	}
	/**
	* Validate the current HTTP request data using a VineJS validator.
	* This method automatically includes request body, files, URL parameters,
	* headers, and cookies in the validation data.
	*
	* @param validator - VineJS validator instance
	* @param options - Optional validation options including custom error reporters and messages
	*
	* @example
	* ```ts
	* const createUserValidator = vine.compile(
	*   vine.object({
	*     email: vine.string().email(),
	*     name: vine.string().minLength(3)
	*   })
	* )
	*
	* const data = await request.validateUsing(createUserValidator, {
	*   errorReporter: () => vine.errors.SimpleErrorReporter,
	*   messagesProvider: customMessages
	* })
	* ```
	*/
	validateUsing(validator, ...[options]) {
		/**
		* Process the validation options
		*/
		const validatorOptions = this.#processValidatorOptions(options);
		/**
		* Data to validate
		*/
		const data = validatorOptions.data || this.#requestData();
		return validator.validate(data, validatorOptions);
	}
	async tryValidateUsing(validator, ...[options]) {
		/**
		* Process the validation options
		*/
		const validatorOptions = this.#processValidatorOptions(options);
		/**
		* Data to validate
		*/
		const data = validatorOptions.data || this.#requestData();
		return validator.tryValidate(data, validatorOptions);
	}
};
//#endregion
//#region modules/http/main.ts
/**
* Bodyparser import is needed to merge types of Request
* class augmented by the bodyparser package
*/
var main_exports = /* @__PURE__ */ __exportAll({ RequestValidator: () => RequestValidator });
import * as import__adonisjs_http_server from "@adonisjs/http-server";
__reExport(main_exports, import__adonisjs_http_server);
//#endregion
export { RequestValidator as n, main_exports as t };
