// @adonisjs/core/build/main-CG-utHjr.js
import { n as __reExport, t as __exportAll } from "./chunk-ZRW5x4aP.js";
import { t as debug_default } from "./debug-DWofTfIX.js";
import { t as configProvider } from "./config_provider-Do9af1ze.js";
import { InvalidArgumentsException } from "@poppinss/utils/exception";
//#region modules/hash/define_config.ts
/**
* Define config for the hash service. This function creates a configuration
* provider that lazily imports and resolves hash drivers when needed.
*
* @param config - Configuration object containing default hasher and list of hashers
* @param config.default - Optional default hasher name (must exist in the list)
* @param config.list - Record of hasher configurations or config providers
*
* @example
* ```ts
* const hashConfig = defineConfig({
*   default: 'scrypt',
*   list: {
*     scrypt: drivers.scrypt({
*       cost: 16384,
*       blockSize: 8,
*       parallelization: 1,
*       saltSize: 16,
*       keyLength: 64,
*     }),
*     bcrypt: drivers.bcrypt({
*       rounds: 10,
*     })
*   }
* })
* ```
*/
function defineConfig(config) {
	/**
	* Hashers list should always be provided
	*/
	if (!config.list) throw new InvalidArgumentsException("Missing \"list\" property in hash config");
	/**
	* The default hasher should be mentioned in the list
	*/
	if (config.default && !config.list[config.default]) throw new InvalidArgumentsException(`Missing "list.${String(config.default)}" in hash config. It is referenced by the "default" property`);
	/**
	* Config provider to lazily import drivers as they are used inside
	* the user application
	*/
	return configProvider.create(async (app) => {
		debug_default("resolving hash config");
		const hashersList = Object.keys(config.list);
		const hashers = {};
		for (let hasherName of hashersList) {
			const hasher = config.list[hasherName];
			if (typeof hasher === "function") hashers[hasherName] = hasher;
			else hashers[hasherName] = await hasher.resolver(app);
		}
		return {
			default: config.default,
			list: hashers
		};
	});
}
/**
* Helpers to configure drivers inside the config file. These functions create
* configuration providers that lazily import and instantiate hash drivers.
*
* - Import happens when you first use the hash module
* - Construction of drivers happens when you first use a driver
*
* @example
* ```ts
* const hashConfig = defineConfig({
*   default: 'bcrypt',
*   list: {
*     bcrypt: drivers.bcrypt({ rounds: 12 }),
*     argon2: drivers.argon2({
*       variant: 'id',
*       memory: 65536,
*       time: 3,
*       parallelism: 4
*     }),
*     scrypt: drivers.scrypt({
*       cost: 16384,
*       blockSize: 8,
*       parallelization: 1
*     })
*   }
* })
* ```
*/
const drivers = {
	argon2: (config) => {
		return configProvider.create(async () => {
			const { Argon } = await import("./modules/hash/drivers/argon.js");
			debug_default("configuring argon driver");
			return () => new Argon(config);
		});
	},
	bcrypt: (config) => {
		return configProvider.create(async () => {
			const { Bcrypt } = await import("./modules/hash/drivers/bcrypt.js");
			debug_default("configuring bcrypt driver");
			return () => new Bcrypt(config);
		});
	},
	scrypt: (config) => {
		return configProvider.create(async () => {
			const { Scrypt } = await import("./modules/hash/drivers/scrypt.js");
			debug_default("configuring scrypt driver");
			return () => new Scrypt(config);
		});
	}
};
//#endregion
//#region modules/hash/main.ts
var main_exports = /* @__PURE__ */ __exportAll({
	defineConfig: () => defineConfig,
	drivers: () => drivers
});
import * as import__adonisjs_hash from "@adonisjs/hash";
__reExport(main_exports, import__adonisjs_hash);
//#endregion
export { defineConfig as n, drivers as r, main_exports as t };
