// @adonisjs/core/build/string-8_8He-HH.js
import prettyHrTime from "pretty-hrtime";
import string from "@poppinss/utils/string";
import he from "he";
import StringBuilder from "@poppinss/utils/string_builder";
//#region src/helpers/string.ts
/**
* Collection of string helpers to transform a string value. This object extends
* the base string utilities from @poppinss/utils with additional AdonisJS-specific
* string manipulation methods.
*
* @example
* // Basic string transformations
* stringHelpers.camelCase('hello_world') // 'helloWorld'
* stringHelpers.snakeCase('HelloWorld')   // 'hello_world'
* stringHelpers.pascalCase('hello world') // 'HelloWorld'
*
* @example
* // HTML escaping and encoding
* stringHelpers.escapeHTML('<script>alert("xss")<\/script>')
* stringHelpers.encodeSymbols('© 2023 AdonisJS')
*/
const stringHelpers = {
	...string,
	toSentence: string.sentence,
	ordinalize: string.ordinal,
	generateRandom: string.random,
	create(value) {
		return new StringBuilder(value);
	},
	prettyHrTime(time, options) {
		return prettyHrTime(time, options);
	},
	isEmpty(value) {
		return value.trim().length === 0;
	},
	escapeHTML(value, options) {
		value = he.escape(value);
		if (options && options.encodeSymbols) value = this.encodeSymbols(value, { allowUnsafeSymbols: true });
		return value;
	},
	encodeSymbols(value, options) {
		return he.encode(value, options);
	}
};
//#endregion
export { stringHelpers as t };
