// @adonisjs/core/build/verification_token-WCrZtUgZ.js
import string from "@poppinss/utils/string";
import base64 from "@poppinss/utils/base64";
import { Secret, safeEqual } from "@poppinss/utils";
import { createHash } from "node:crypto";
//#region src/helpers/verification_token.ts
/**
* Verification token class can be used to create tokens publicly
* shareable tokens while storing the token hash within the database.
*
* This class is used by the Auth and the Persona packages to manage
* tokens for authentication and authorization purposes.
*
* @example
* class UserToken extends VerificationToken {
*   constructor(user: User, secret: Secret<string>) {
*     super()
*     this.tokenableId = user.id
*     this.computeValue(secret)
*   }
* }
*/
var VerificationToken = class {
	/**
	* Decodes a publicly shared token and return the series
	* and the token value from it.
	*
	* Returns null when unable to decode the token because of
	* invalid format or encoding.
	*
	* @param value - The token string to decode
	*/
	static decode(value) {
		/**
		* Ensure value is a string and starts with the prefix.
		*/
		if (typeof value !== "string") return null;
		/**
		* Remove prefix from the rest of the token.
		*/
		if (!value) return null;
		const [identifier, ...tokenValue] = value.split(".");
		if (!identifier || tokenValue.length === 0) return null;
		const decodedIdentifier = base64.urlDecode(identifier);
		const decodedSecret = base64.urlDecode(tokenValue.join("."));
		if (!decodedIdentifier || !decodedSecret) return null;
		return {
			identifier: decodedIdentifier,
			secret: new Secret(decodedSecret)
		};
	}
	/**
	* Creates a transient token that can be shared with the persistence
	* layer.
	*
	* @param userId - The user ID for whom the token is being created
	* @param size - The size of the random token seed
	* @param expiresIn - Token expiration time (string like '2h' or number in seconds)
	*/
	static createTransientToken(userId, size, expiresIn) {
		const expiresAt = /* @__PURE__ */ new Date();
		expiresAt.setSeconds(expiresAt.getSeconds() + string.seconds.parse(expiresIn));
		return {
			userId,
			expiresAt,
			...this.seed(size)
		};
	}
	/**
	* Creates a secret opaque token and its hash.
	*
	* @param size - The length of the random token to generate
	*/
	static seed(size) {
		const secret = new Secret(string.random(size));
		return {
			secret,
			hash: createHash("sha256").update(secret.release()).digest("hex")
		};
	}
	/**
	* Compute the value property using the given secret. You can
	* get secret via the static "createTransientToken" method.
	*
	* @param secret - The secret value to compute the public token value from
	*/
	computeValue(secret) {
		this.value = new Secret(`${base64.urlEncode(String(this.identifier))}.${base64.urlEncode(secret.release())}`);
	}
	/**
	* Check if the token has been expired. Verifies
	* the "expiresAt" timestamp with the current
	* date.
	*/
	isExpired() {
		return this.expiresAt < /* @__PURE__ */ new Date();
	}
	/**
	* Verifies the value of a token against the pre-defined hash
	*
	* @param secret - The secret to verify against the stored hash
	*/
	verify(secret) {
		const newHash = createHash("sha256").update(secret.release()).digest("hex");
		return safeEqual(this.hash, newHash);
	}
};
//#endregion
export { VerificationToken as t };
