// @adonisjs/core/build/commands/list/routes.js
import { o as args, p as BaseCommand, s as cliHelpers, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
import stringWidth from "string-width";
import { middlewareInfo, routeInfo } from "@adonisjs/http-server/helpers";
//#region src/cli_formatters/routes_list.ts
/**
* Routes list formatter is used to format the routes to JSON or an ANSI string
* with pretty output.
*
* The decisions of colors, padding, alignment are all handled by the lists formatter
*
* @example
* const formatter = new RoutesListFormatter(router, ui, {
*   displayHeadRoutes: false,
*   maxPrettyPrintWidth: 120
* }, {
*   match: 'api',
*   middleware: ['auth']
* })
*
* const ansiOutput = await formatter.formatAsAnsiList()
*/
var RoutesListFormatter = class {
	/**
	* Router instance containing all registered routes
	*/
	#router;
	/**
	* Colors utility for ANSI formatting
	*/
	#colors;
	/**
	* Table utility for creating formatted tables
	*/
	#table;
	/**
	* Options for printing routes
	*/
	#options;
	/**
	* Filters to apply when finding routes
	*/
	#filters;
	/**
	* Creates a new instance of the routes list formatter
	*
	* @param router - Router instance containing routes to format
	* @param ui - UI primitives for colors and table formatting
	* @param options - Display options for route formatting
	* @param filters - Filters to apply when displaying routes
	*/
	constructor(router, ui, options, filters) {
		this.#router = router;
		this.#colors = ui.colors;
		this.#table = ui.table;
		this.#filters = filters;
		this.#options = options;
		this.#router.commit();
	}
	/**
	* Test if a route clears the applied filters based on middleware, name, pattern, and handler.
	*
	* @param route - The serialized route to test against filters
	*/
	#isAllowedByFilters(route) {
		let allowRoute = true;
		/**
		* Check if the route is allowed by applying the middleware
		* filter
		*/
		if (this.#filters.middleware) allowRoute = this.#filters.middleware.every((name) => {
			if (name === "*") return route.middleware.length > 0;
			return route.middleware.find((middleware) => middleware.name === name);
		});
		/**
		* Check if the route has any or the ignored middleware. If yes, do not
		* display the route
		*/
		if (allowRoute && this.#filters.ignoreMiddleware) allowRoute = this.#filters.ignoreMiddleware.every((name) => {
			if (name === "*") return route.middleware.length === 0;
			return !route.middleware.find((middleware) => middleware.name === name);
		});
		/**
		* No more filters to be applied
		*/
		if (!this.#filters.match) return allowRoute;
		/**
		* Check if the route name has the match keyword
		*/
		if (route.name.includes(this.#filters.match)) return true;
		/**
		* Check if the route pattern has the match keyword
		*/
		if (route.pattern.includes(this.#filters.match)) return true;
		/**
		* Check if the route handler has the match keyword
		*/
		if (route.handler.type === "controller" ? route.handler.moduleNameOrPath.includes(this.#filters.match) : route.handler.name.includes(this.#filters.match)) return true;
		/**
		* Disallow route
		*/
		return false;
	}
	/**
	* Serializes routes JSON to an object that can be used for pretty printing.
	* Converts RouteJSON into a format suitable for display and filtering.
	*
	* @param route - The route JSON object to serialize
	*/
	async #serializeRoute(route) {
		let methods = route.methods;
		if (!this.#options.displayHeadRoutes) methods = methods.filter((method) => method !== "HEAD");
		const middlewareList = await Promise.all([...route.middleware.all()].map((middleware) => {
			return middlewareInfo(middleware);
		}));
		return {
			name: route.name || "",
			pattern: route.pattern,
			methods,
			handler: await routeInfo(route),
			middleware: middlewareList.filter((info) => info.type !== "global")
		};
	}
	/**
	* Formats the route method for the ansi list and table with dim styling.
	*
	* @param method - The HTTP method to format (GET, POST, etc.)
	*/
	#formatRouteMethod(method) {
		return this.#colors.dim(method);
	}
	/**
	* Formats route pattern for the ansi list and table with colored parameters and route name.
	*
	* @param route - The serialized route containing pattern and name information
	*/
	#formatRoutePattern(route) {
		const pattern = this.#router.parsePattern(route.pattern).map((token) => {
			if (token.type === 1) return this.#colors.yellow(`:${token.val}`);
			if (token.type === 3) return this.#colors.yellow(`:${token.val}?`);
			if (token.type === 2) return this.#colors.red(token.val);
			return token.val;
		}).join("/");
		return `${pattern === "/" ? pattern : `/${pattern}`}${route.name ? ` ${this.#colors.dim(`(${route.name})`)}` : ""} `;
	}
	/**
	* Formats controller name for the ansi list and table with cyan coloring.
	*
	* @param route - The serialized route containing handler information
	*/
	#formatControllerName(route) {
		return route.handler.type === "controller" ? ` ${this.#colors.cyan(route.handler.moduleNameOrPath)}.` : "";
	}
	/**
	* Formats action name for the ansi list and table with cyan coloring and arguments.
	*
	* @param route - The serialized route containing handler information
	*/
	#formatAction(route) {
		if (route.handler.type === "controller") return `${this.#colors.cyan(route.handler.method)}`;
		const functionName = ` ${this.#colors.cyan(route.handler.name)}`;
		if (route.handler.args) return ` ${functionName}${this.#colors.dim(`(${route.handler.args})`)}`;
		return functionName;
	}
	/**
	* Formats route middleware for the ansi list and table with optional compacting.
	*
	* @param route - The serialized route containing middleware information
	* @param mode - Display mode: 'normal' shows all middleware, 'compact' truncates long lists
	*/
	#formatMiddleware(route, mode = "normal") {
		if (mode === "compact" && route.middleware.length > 3) {
			const firstMiddleware = route.middleware[0].name;
			const secondMiddleware = route.middleware[1].name;
			const diff = route.middleware.length - 2;
			return this.#colors.dim(`${firstMiddleware}, ${secondMiddleware}, and ${diff} more`);
		}
		return this.#colors.dim(`${route.middleware.map((one) => one.name).filter((one) => one).join(", ")}`);
	}
	/**
	* Formatting the domain headling to be in green color with
	* dots around it
	*/
	#formatDomainHeadline(domain) {
		if (domain !== "root") return cliHelpers.justify([`${this.#colors.dim("..")} ${this.#colors.green(domain)} `], {
			maxWidth: this.#options.maxPrettyPrintWidth || cliHelpers.TERMINAL_SIZE,
			paddingChar: this.#colors.dim(".")
		})[0];
		return "";
	}
	/**
	* Justify the ansi list
	*/
	#justifyListTables(tables) {
		return tables.map((table) => {
			/**
			* Formatting methods
			*/
			const methods = table.rows.map((columns) => columns[0]);
			const largestMethodsLength = Math.max(...methods.map((method) => stringWidth(method)));
			const formattedMethods = cliHelpers.justify(methods, { maxWidth: largestMethodsLength });
			/**
			* Formatting patterns
			*/
			const patterns = table.rows.map((columns) => columns[1]);
			const largestPatternLength = Math.max(...patterns.map((pattern) => stringWidth(pattern)));
			const formattedPatterns = cliHelpers.justify(patterns, {
				maxWidth: largestPatternLength,
				paddingChar: this.#colors.dim(".")
			});
			/**
			* Formatting middleware to be right aligned
			*/
			const middleware = table.rows.map((columns) => columns[3]);
			const largestMiddlewareLength = Math.max(...middleware.map((one) => stringWidth(one)));
			const formattedMiddleware = cliHelpers.justify(middleware, {
				maxWidth: largestMiddlewareLength,
				align: "right",
				paddingChar: " "
			});
			/**
			* Formatting controllers to be right aligned and take all the remaining
			* space after printing route method, pattern and middleware.
			*/
			const controllers = table.rows.map((columns) => columns[2]);
			const largestControllerLength = (this.#options.maxPrettyPrintWidth || cliHelpers.TERMINAL_SIZE) - (largestPatternLength + largestMethodsLength + largestMiddlewareLength);
			const formattedControllers = cliHelpers.truncate(cliHelpers.justify(controllers, {
				maxWidth: largestControllerLength,
				align: "right",
				paddingChar: this.#colors.dim(".")
			}), { maxWidth: largestControllerLength });
			return {
				heading: table.heading,
				rows: formattedMethods.reduce((result, method, index) => {
					result.push(`${method}${formattedPatterns[index]}${formattedControllers[index]}${formattedMiddleware[index]}`);
					return result;
				}, [])
			};
		});
	}
	/**
	* Formats middleware info into a compact string representation.
	* Named middleware with args becomes "name:args", closures use their name as-is.
	*/
	#formatMiddlewareAsString(middleware) {
		if (middleware.type === "named" && middleware.args) return `${middleware.name}:${middleware.args}`;
		return middleware.name;
	}
	/**
	* Formats routes as JSONL (one JSON object per line). Each line is a
	* self-contained route object with flattened domain, simplified handler,
	* and middleware as a string array. Optimized for machine consumption
	* by AI agents and CLI tools.
	*/
	async formatAsJSONL() {
		const routes = this.#router.toJSON();
		const domains = Object.keys(routes);
		const lines = [];
		for (let domain of domains) for (let route of routes[domain]) {
			const serializedRoute = await this.#serializeRoute(route);
			if (!this.#isAllowedByFilters(serializedRoute)) continue;
			const handler = serializedRoute.handler.type === "controller" ? {
				type: "controller",
				module: serializedRoute.handler.moduleNameOrPath,
				method: serializedRoute.handler.method
			} : {
				type: serializedRoute.handler.name === "redirectsToRoute" ? "redirect" : "closure",
				name: serializedRoute.handler.name,
				...serializedRoute.handler.args ? { args: serializedRoute.handler.args } : {}
			};
			const middleware = serializedRoute.middleware.map((m) => this.#formatMiddlewareAsString(m)).filter((m) => !!m);
			for (let method of serializedRoute.methods) {
				const entry = {
					method,
					pattern: serializedRoute.pattern,
					handler
				};
				if (serializedRoute.name) entry.name = serializedRoute.name;
				if (domain !== "root") entry.domain = domain;
				if (middleware.length > 0) entry.middleware = middleware;
				lines.push(JSON.stringify(entry));
			}
		}
		return lines;
	}
	/**
	* Formats routes as an array of objects. Routes are grouped by
	* domain.
	*/
	async formatAsJSON() {
		const routes = this.#router.toJSON();
		const domains = Object.keys(routes);
		let routesJSON = [];
		for (let domain of domains) {
			const domainRoutes = await Promise.all(routes[domain].map((route) => this.#serializeRoute(route)));
			routesJSON.push({
				domain,
				routes: domainRoutes.filter((route) => this.#isAllowedByFilters(route))
			});
		}
		return routesJSON;
	}
	/**
	* Format routes to ansi list of tables. Each domain has its own table
	* with heading and rows. Each row has colums with colors and spacing
	* around them.
	*/
	async formatAsAnsiList() {
		const routes = this.#router.toJSON();
		const domains = Object.keys(routes);
		const tables = [];
		for (let domain of domains) {
			const list = {
				heading: this.#formatDomainHeadline(domain),
				rows: [[
					this.#colors.dim("METHOD"),
					` ${this.#colors.dim("ROUTE")} `,
					` ${this.#colors.dim("HANDLER")}`,
					` ${this.#colors.dim("MIDDLEWARE")}`
				]]
			};
			/**
			* Computing table rows. Each route+method will have its
			* own row
			*/
			for (let route of routes[domain]) {
				const serializedRoute = await this.#serializeRoute(route);
				if (this.#isAllowedByFilters(serializedRoute)) serializedRoute.methods.forEach((method) => {
					list.rows.push([
						this.#formatRouteMethod(method),
						` ${this.#formatRoutePattern(serializedRoute)}`,
						`${this.#formatControllerName(serializedRoute)}${this.#formatAction(serializedRoute)}`,
						` ${this.#formatMiddleware(serializedRoute, "compact")}`
					]);
				});
			}
			tables.push(list);
		}
		return this.#justifyListTables(tables);
	}
	/**
	* Format routes to ansi tables. Each domain has its own table
	* with heading and rows. Each row has colums with colors and spacing
	* around them.
	*/
	async formatAsAnsiTable() {
		const routes = this.#router.toJSON();
		const domains = Object.keys(routes);
		const tables = [];
		for (let domain of domains) {
			const list = {
				heading: this.#formatDomainHeadline(domain),
				table: this.#table().fullWidth().fluidColumnIndex(2).head([
					this.#colors.dim("METHOD"),
					this.#colors.dim("ROUTE"),
					{
						hAlign: "right",
						content: this.#colors.dim("HANDLER")
					},
					{
						content: this.#colors.dim("MIDDLEWARE"),
						hAlign: "right"
					}
				])
			};
			/**
			* Computing table rows. Each route+method will have its
			* own row
			*/
			for (let route of routes[domain]) {
				const serializedRoute = await this.#serializeRoute(route);
				if (this.#isAllowedByFilters(serializedRoute)) serializedRoute.methods.forEach((method) => {
					list.table.row([
						this.#formatRouteMethod(method),
						this.#formatRoutePattern(serializedRoute),
						{
							content: `${this.#formatControllerName(serializedRoute)}${this.#formatAction(serializedRoute)}`,
							hAlign: "right"
						},
						{
							content: this.#formatMiddleware(serializedRoute),
							hAlign: "right"
						}
					]);
				});
			}
			tables.push(list);
		}
		return tables;
	}
};
//#endregion
//#region commands/list/routes.ts
/**
* Command to display a list of all registered routes in the application.
* Supports filtering by keywords, middleware, and output formatting options.
* Routes can be displayed as a formatted list, table, or JSON.
*
* @example
* ```
* ace list:routes
* ace list:routes user
* ace list:routes --middleware=auth
* ace list:routes --ignore-middleware=guest
* ace list:routes --json
* ace list:routes --table
* ```
*/
var ListRoutes = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "list:routes";
	/**
	* The command description
	*/
	static description = "List all registered routes with their HTTP methods, URL patterns, handlers, and middleware";
	/**
	* Command options configuration.
	* Requires the application to be started so routes are loaded.
	*/
	static options = { startApp: true };
	/**
	* Execute the command to list application routes.
	* Creates a formatter with the specified filters and outputs routes
	* in the requested format (JSON, table, or formatted list).
	*/
	async run() {
		const formatter = new RoutesListFormatter(await this.app.container.make("router"), this.ui, {}, {
			ignoreMiddleware: this.ignoreMiddleware,
			middleware: this.middleware,
			match: this.match
		});
		/**
		* Display as JSONL (one JSON object per line).
		* Auto-selected when running inside an AI agent and no
		* explicit format flag is provided.
		*/
		if (this.jsonl || !this.json && !this.table && this.app.runningInAIAgent) {
			const lines = await formatter.formatAsJSONL();
			for (const line of lines) this.logger.log(line);
			return;
		}
		/**
		* Display as JSON
		*/
		if (this.json) {
			this.logger.log(JSON.stringify(await formatter.formatAsJSON(), null, 2));
			return;
		}
		/**
		* Display as a standard table
		*/
		if (this.table) {
			(await formatter.formatAsAnsiTable()).forEach((table) => {
				this.logger.log("");
				if (table.heading) {
					this.logger.log(table.heading);
					this.logger.log("");
				}
				table.table.render();
			});
			return;
		}
		(await formatter.formatAsAnsiList()).forEach((item) => {
			this.logger.log("");
			if (item.heading) {
				this.logger.log(item.heading);
				this.logger.log("");
			}
			this.logger.log(item.rows.join("\n"));
		});
	}
};
__decorate([args.string({
	description: "Find routes matching the given keyword. Route name, pattern and controller name will be searched against the keyword",
	required: false
})], ListRoutes.prototype, "match", void 0);
__decorate([flags.array({ description: "View routes that includes all the mentioned middleware names. Use * to see routes that are using one or more middleware" })], ListRoutes.prototype, "middleware", void 0);
__decorate([flags.array({ description: "View routes that does not include all the mentioned middleware names. Use * to see routes that are using zero middleware" })], ListRoutes.prototype, "ignoreMiddleware", void 0);
__decorate([flags.boolean({ description: "Get routes list as a JSON string" })], ListRoutes.prototype, "json", void 0);
__decorate([flags.boolean({ description: "View list of routes as a table" })], ListRoutes.prototype, "table", void 0);
__decorate([flags.boolean({ description: "Get routes as JSONL, one JSON object per line (optimized for AI agents)" })], ListRoutes.prototype, "jsonl", void 0);
//#endregion
export { ListRoutes as default };
