// @adonisjs/core/build/providers/edge_provider.js
import { t as main_exports } from "../main-BDd3QiJF.js";
import { t as pluginEdgeDumper } from "../edge-BG5ZYp6Z.js";
import edge from "edge.js";
//#region providers/edge_provider.ts
/**
* The Edge service provider configures Edge to work within
* an AdonisJS application environment
*
* This provider integrates EdgeJS template engine with AdonisJS by:
* - Mounting the views directory
* - Configuring template caching for production
* - Adding global helpers for route generation
* - Creating isolated renderer instances for HTTP contexts
* - Adding render macro to BriskRoute for template rendering
*
* @example
* const provider = new EdgeServiceProvider(app)
* await provider.boot()
*/
var EdgeServiceProvider = class {
	/**
	* Edge service provider constructor
	*
	* Sets the usingEdgeJS flag to true to indicate EdgeJS is being used.
	*
	* @param app - The application service instance
	*/
	constructor(app) {
		this.app = app;
		this.app.usingEdgeJS = true;
	}
	/**
	* Bridge AdonisJS and Edge
	*
	* Configures EdgeJS integration by:
	* - Setting up template mounting and caching
	* - Defining global helpers (route, signedRoute, app, config)
	* - Adding view getter to HttpContext for isolated rendering
	* - Adding render macro to BriskRoute
	* - Registering dumper plugin
	*
	* @example
	* await provider.boot()
	* // Now edge templates can use {{ route('home') }} helper
	*/
	async boot() {
		const app = this.app;
		const qs = new main_exports.Qs(app.config.get("app.http.qs", {}));
		const router = await this.app.container.make("router");
		const dumper = await this.app.container.make("dumper");
		/**
		* Resolves configuration values for Edge templates
		*
		* Provides access to application configuration within templates.
		* Includes a 'has' method to check for config key existence.
		*
		* @param key - The configuration key to retrieve
		* @param defaultValue - Optional default value if key doesn't exist
		*/
		function edgeConfigResolver(key, defaultValue) {
			return app.config.get(key, defaultValue);
		}
		edgeConfigResolver.has = function(key) {
			return app.config.has(key);
		};
		/**
		* Generates client-side route definitions for frontend use
		*
		* Transforms router definitions into a serializable format that
		* can be used in client-side JavaScript for route generation.
		* Only includes named routes.
		*/
		function clientRoutes() {
			const routes = router.toJSON();
			return Object.keys(routes).reduce((result, domain) => {
				result[domain] = routes[domain].reduce((routesResult, route) => {
					if (!route.name) return routesResult;
					routesResult.push({
						domain: route.domain,
						methods: route.methods,
						pattern: route.pattern,
						tokens: route.tokens,
						name: route.name
					});
					return routesResult;
				}, []);
				return result;
			}, {});
		}
		/**
		* Mount the default disk
		*/
		edge.mount(app.viewsPath());
		/**
		* Cache templates in production
		*/
		edge.configure({ cache: app.inProduction });
		/**
		* Define Edge global helpers
		* @deprecated
		*/
		edge.global("route", function(...args) {
			return router.makeUrl(...args);
		});
		edge.global("signedRoute", function(...args) {
			return router.makeSignedUrl(...args);
		});
		edge.global("app", app);
		edge.global("config", edgeConfigResolver);
		edge.global("routes", function() {
			return clientRoutes();
		});
		edge.global("routesJSON", function() {
			return JSON.stringify(clientRoutes());
		});
		/**
		* Route helpers
		*/
		edge.global("urlFor", function(...args) {
			return router.urlBuilder.urlFor(...args);
		});
		edge.global("signedUrlFor", function(...args) {
			return router.urlBuilder.signedUrlFor(...args);
		});
		/**
		* Sharing qs parser with templates
		*/
		edge.global("qs", qs);
		edge.global("formAttributes", function(route, params, options) {
			const matchingRoute = router.findOrFail(route);
			/**
			* Normalize method and keep a reference to the original method
			*/
			options = options ?? {};
			let method = matchingRoute.methods[0].toUpperCase();
			const original = method;
			/**
			* In case of HEAD, we must use the GET method
			*/
			if (method === "HEAD") method = "GET";
			/**
			* If method if not GET and POST, then use the querystring _method
			* to and force update the method to "POST"
			*/
			if (method !== "GET" && method !== "POST") {
				method = "POST";
				options = {
					...options,
					qs: {
						_method: original,
						...options.qs
					}
				};
			}
			const { action } = router.urlBuilder.urlFor.method(original, route, params, options).form;
			return {
				action,
				method
			};
		});
		/**
		* Creating a isolated instance of edge renderer
		*/
		main_exports.HttpContext.getter("view", function() {
			return edge.createRenderer().share({ request: this.request });
		}, true);
		/**
		* Adding brisk route to render templates without an
		* explicit handler
		*/
		main_exports.BriskRoute.macro("render", function(template, data) {
			function rendersTemplate({ view }) {
				return view.render(template, data);
			}
			Object.defineProperty(rendersTemplate, "listArgs", {
				value: template,
				writable: false
			});
			return this.setHandler(rendersTemplate);
		});
		edge.use(pluginEdgeDumper(dumper));
	}
};
//#endregion
export { EdgeServiceProvider as default };
