// @vinejs/vine/build/simple_error_reporter-EQTcwKJ9.js
import delve from "dlv";
import dayjs from "dayjs";
import isIP from "validator/lib/isIP.js";
import isJWT from "validator/lib/isJWT.js";
import isURL from "validator/lib/isURL.js";
import isSlug from "validator/lib/isSlug.js";
import isIBAN from "validator/lib/isIBAN.js";
import isUUID from "validator/lib/isUUID.js";
import isAscii from "validator/lib/isAscii.js";
import isEmail from "validator/lib/isEmail.js";
import isAlpha from "validator/lib/isAlpha.js";
import isLatLong from "validator/lib/isLatLong.js";
import isDecimal from "validator/lib/isDecimal.js";
import isHexColor from "validator/lib/isHexColor.js";
import isCreditCard from "validator/lib/isCreditCard.js";
import isSameOrAfter from "dayjs/plugin/isSameOrAfter.js";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore.js";
import isAlphanumeric from "validator/lib/isAlphanumeric.js";
import isPassportNumber from "validator/lib/isPassportNumber.js";
import isVAT from "validator/lib/isVAT.js";
import customParseFormat from "dayjs/plugin/customParseFormat.js";
import isPostalCode, { locales } from "validator/lib/isPostalCode.js";
import isMobilePhone, { locales as locales$1 } from "validator/lib/isMobilePhone.js";
//#region \0rolldown/runtime.js
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
//#endregion
//#region src/vine/helpers.ts
/**
* Values that are considered true in HTML form context.
* Includes boolean true, number 1, string '1', 'true', and 'on' (for checkboxes).
*/
const BOOLEAN_POSITIVES = [
	"1",
	1,
	"true",
	true,
	"on"
];
/**
* Values that are considered false in HTML form context.
* Includes boolean false, number 0, string '0', and 'false'.
*/
const BOOLEAN_NEGATIVES = [
	"0",
	0,
	"false",
	false
];
/**
* Default date formats used when no specific format is provided.
* Supports date-only (YYYY-MM-DD) and datetime (YYYY-MM-DD HH:mm:ss) formats.
*/
const DEFAULT_DATE_FORMATS = ["YYYY-MM-DD", "YYYY-MM-DD HH:mm:ss"];
/**
* Regular expression for validating ULID (Universally Unique Lexicographically Sortable Identifier) format.
* ULIDs are 26 characters long using Crockford's Base32 encoding.
*/
const ULID = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/;
dayjs.extend(customParseFormat);
dayjs.extend(isSameOrAfter);
dayjs.extend(isSameOrBefore);
/**
* Collection of utility helpers used across the Vine validation library.
* These helpers provide type-checking, coercion, and validation functions
* that are optimized for HTML form data handling.
*
* @example
* helpers.isString(value) // Type guard for strings
* helpers.asBoolean(value) // Convert to boolean with HTML form semantics
* helpers.isEmail(email) // Validate email addresses
*/
const helpers = {
	exists(value) {
		return value !== null && value !== void 0;
	},
	isMissing(value) {
		return !this.exists(value);
	},
	isTrue(value) {
		return BOOLEAN_POSITIVES.includes(value);
	},
	isFalse(value) {
		return BOOLEAN_NEGATIVES.includes(value);
	},
	isString(value) {
		return typeof value === "string";
	},
	isObject(value) {
		return !!(value && typeof value === "object" && !Array.isArray(value));
	},
	hasKeys(value, keys) {
		for (let key of keys) if (key in value === false) return false;
		return true;
	},
	isArray(value) {
		return Array.isArray(value);
	},
	isNumeric(value) {
		return !Number.isNaN(Number(value));
	},
	asNumber(value) {
		return value === null ? NaN : Number(value);
	},
	asBoolean(value) {
		if (this.isTrue(value)) return true;
		if (this.isFalse(value)) return false;
		return null;
	},
	asDayJS(value, format) {
		let isTimestampAllowed = false;
		let isISOAllowed = false;
		let formats = format || DEFAULT_DATE_FORMATS;
		/**
		* DayJS mutates the formats property under the hood. Therefore
		* we have to create a shallow clone before passing formats.
		*
		* https://github.com/iamkun/dayjs/issues/2136
		*/
		if (Array.isArray(formats)) {
			formats = [...formats];
			isTimestampAllowed = formats.includes("x");
			isISOAllowed = formats.includes("iso8601");
		} else if (typeof formats !== "string") {
			formats = { ...formats };
			isTimestampAllowed = formats.format === "x";
			isISOAllowed = formats.format === "iso";
		}
		const valueAsNumber = isTimestampAllowed ? helpers.asNumber(value) : value;
		let dateTime;
		/**
		* The timestamp validation does not work with formats array.
		* Therefore we validate is separately without passing any
		* formats.
		*
		* Otherwise we parse the date with formats
		*/
		if (isTimestampAllowed && !Number.isNaN(valueAsNumber)) dateTime = dayjs(valueAsNumber);
		else dateTime = dayjs(value, formats, true);
		/**
		* If datetime is invalid and the ISO format is allowed,
		* then we reattempt to parse the date without formats
		*/
		if (!dateTime.isValid() && isISOAllowed) dateTime = dayjs(value);
		return {
			dateTime,
			formats
		};
	},
	compareValues(inputValue, expectedValue) {
		let input = inputValue;
		/**
		* Normalize the field value based on the expected value type
		*/
		if (typeof expectedValue === "boolean") input = this.asBoolean(inputValue);
		else if (typeof expectedValue === "number") input = this.asNumber(inputValue);
		/**
		* Perform comparison and return results
		*/
		return {
			isEqual: input === expectedValue,
			casted: input
		};
	},
	isEmail: isEmail.default,
	isURL: isURL.default,
	isAlpha: isAlpha.default,
	isAlphaNumeric: isAlphanumeric.default,
	isIP: isIP.default,
	isUUID: isUUID.default,
	isAscii: isAscii.default,
	isCreditCard: isCreditCard.default,
	isIBAN: isIBAN.default,
	isJWT: isJWT.default,
	isLatLong: isLatLong.default,
	isMobilePhone: isMobilePhone.default,
	isPassportNumber: isPassportNumber.default,
	isVAT: isVAT.default,
	isPostalCode: isPostalCode.default,
	isSlug: isSlug.default,
	isDecimal: isDecimal.default,
	mobileLocales: locales$1,
	postalCountryCodes: locales,
	passportCountryCodes: [
		"AM",
		"AR",
		"AT",
		"AU",
		"AZ",
		"BE",
		"BG",
		"BR",
		"BY",
		"CA",
		"CH",
		"CY",
		"CZ",
		"DE",
		"DK",
		"DZ",
		"ES",
		"FI",
		"FR",
		"GB",
		"GR",
		"HR",
		"HU",
		"IE",
		"IN",
		"ID",
		"IR",
		"IS",
		"IT",
		"JM",
		"JP",
		"KR",
		"KZ",
		"LI",
		"LT",
		"LU",
		"LV",
		"LY",
		"MT",
		"MZ",
		"MY",
		"MX",
		"NL",
		"NZ",
		"PH",
		"PK",
		"PL",
		"PT",
		"RO",
		"RU",
		"SE",
		"SL",
		"SK",
		"TH",
		"TR",
		"UA",
		"US"
	],
	isULID(value) {
		if (typeof value !== "string") return false;
		if (value[0] > "7") return false;
		return ULID.test(value);
	},
	isHexColor: (value) => {
		if (!value.startsWith("#")) return false;
		return isHexColor.default(value);
	},
	isActiveURL: async (url) => {
		const { resolve4, resolve6 } = await import("node:dns/promises");
		try {
			const { hostname } = new URL(url);
			if ((await resolve6(hostname)).length) return true;
			else return (await resolve4(hostname)).length > 0;
		} catch {
			return false;
		}
	},
	isDistinct: (dataSet, fields) => {
		const uniqueItems = /* @__PURE__ */ new Set();
		/**
		* Check for duplicates when no fields are provided
		*/
		if (!fields) {
			for (let item of dataSet) if (helpers.exists(item)) if (uniqueItems.has(item)) return false;
			else uniqueItems.add(item);
			return true;
		}
		/**
		* Checking for duplicates when one or more fields are mentioned
		*/
		const fieldsList = Array.isArray(fields) ? fields : [fields];
		for (let item of dataSet)
 /**
		* Only process item, if it is an object and has all the fields
		* required for uniqueness check
		*/
		if (helpers.isObject(item) && helpers.hasKeys(item, fieldsList)) {
			const element = fieldsList.map((field) => item[field]).join("_");
			if (uniqueItems.has(element)) return false;
			else uniqueItems.add(element);
		}
		return true;
	},
	getNestedValue(key, field) {
		if (key.indexOf(".") > -1) return delve(field.data, key);
		return field.parent[key];
	},
	optional(props) {
		const result = {};
		for (const name of Object.keys(props)) {
			let field = props[name].clone();
			if ("optional" in field && typeof field.optional === "function") field = field.optional();
			result[name] = field;
		}
		return result;
	}
};
//#endregion
//#region src/messages_provider/simple_messages_provider.ts
/**
* Default messages provider that performs message lookup using key-value pairs.
* Provides internationalization and custom error message support for validation.
* Supports field-specific messages, wildcard patterns, and placeholder interpolation.
*
* @example
* const provider = new SimpleMessagesProvider({
*   'required': 'The {{ field }} field is required',
*   'email': 'The {{ field }} field must be a valid email',
*   'user.email.required': 'Email address is required'
* }, {
*   'user.email': 'Email Address'
* })
*/
var SimpleMessagesProvider = class {
	/** Collection of validation error message templates */
	#messages;
	/** Mapping of field paths to human-readable field names */
	#fields;
	/**
	* Creates a new SimpleMessagesProvider instance.
	*
	* @param messages - Map of validation rule names to error message templates
	* @param fields - Optional map of field paths to human-readable names
	*
	* @example
	* new SimpleMessagesProvider({
	*   'required': 'The {{ field }} field is required'
	* }, {
	*   'user_name': 'Username'
	* })
	*/
	constructor(messages, fields) {
		this.#messages = messages;
		this.#fields = fields || {};
	}
	/**
	* Interpolates placeholders within error messages using mustache-like syntax.
	* Supports nested property access using dot notation.
	*
	* @param message - The message template with placeholders
	* @param data - Data object containing values for interpolation
	* @returns Interpolated message string
	*
	* @example
	* interpolate('The {{ field }} must be at least {{ min }} characters', {
	*   field: 'username',
	*   min: 5
	* })
	* // Returns: "The username must be at least 5 characters"
	*/
	#interpolate(message, data) {
		if (!message.includes("{{")) return message;
		return message.replace(/(\\)?{{(.*?)}}/g, (_, __, key) => {
			const tokens = key.trim().split(".");
			let output = data;
			while (tokens.length) {
				if (output === null || typeof output !== "object") return;
				const token = tokens.shift();
				output = Object.hasOwn(output, token) ? output[token] : void 0;
			}
			return output;
		});
	}
	/**
	* Returns a validation error message for a given field and rule.
	* Uses a priority-based lookup system to find the most specific message.
	*
	* Priority order:
	* 1. Field-specific messages (e.g., 'user.email.required')
	* 2. Wildcard path messages (e.g., '*.email.required')
	* 3. Generic rule messages (e.g., 'required')
	* 4. Fallback to raw message
	*
	* @param rawMessage - The default raw message from the validation rule
	* @param rule - The name of the validation rule that failed
	* @param field - Context information about the field being validated
	* @param args - Additional arguments to interpolate into the message
	* @returns Formatted and interpolated error message
	*
	* @example
	* // With field-specific message
	* provider.getMessage('Required', 'required', fieldContext, {})
	* // Returns: "Email address is required" (if field-specific message exists)
	*
	* // With rule arguments
	* provider.getMessage('Min length', 'minLength', fieldContext, { min: 5 })
	* // Returns: "The username must be at least 5 characters"
	*/
	getMessage(rawMessage, rule, field, args) {
		const fieldName = this.#fields[field.name] || field.name;
		/**
		* 1st priority: Field-specific messages (highest specificity)
		* Example: 'user.email.required'
		*/
		const fieldMessage = this.#messages[`${field.getFieldPath()}.${rule}`];
		if (fieldMessage) return this.#interpolate(fieldMessage, {
			field: fieldName,
			...args
		});
		/**
		* 2nd priority: Wildcard path messages (medium specificity)
		* Example: '*.email.required'
		*/
		const wildcardMessage = this.#messages[`${field.wildCardPath}.${rule}`];
		if (wildcardMessage) return this.#interpolate(wildcardMessage, {
			field: fieldName,
			...args
		});
		/**
		* 3rd priority: Generic rule messages (low specificity)
		* Example: 'required'
		*/
		const ruleMessage = this.#messages[rule];
		if (ruleMessage) return this.#interpolate(ruleMessage, {
			field: fieldName,
			...args
		});
		/**
		* Fallback: Use the raw message provided by the validation rule
		*/
		return this.#interpolate(rawMessage, {
			field: fieldName,
			...args
		});
	}
	/**
	* Serializes the messages provider to a JSON-compatible object.
	* Useful for debugging, logging, or transferring configuration.
	*
	* @returns Object containing messages and field mappings
	*
	* @example
	* const config = provider.toJSON()
	* console.log(config)
	* // { messages: { required: "..." }, fields: { user_name: "Username" } }
	*/
	toJSON() {
		return {
			messages: this.#messages,
			fields: this.#fields
		};
	}
};
//#endregion
//#region src/errors/validation_error.ts
/**
* ValidationError is a specialized Error class that represents validation
* failures. It contains structured error messages and HTTP status information
* for easy integration with web frameworks.
*
* @example
* try {
*   await vine.validate({ schema, data })
* } catch (error) {
*   if (error instanceof ValidationError) {
*     console.log(error.messages) // Structured validation errors
*     console.log(error.status)   // HTTP 422
*   }
* }
*/
var ValidationError = class extends Error {
	/**
	* HTTP status code for the validation error (422 Unprocessable Entity)
	*/
	status = 422;
	/**
	* Internal error code for programmatic error handling
	*/
	code = "E_VALIDATION_ERROR";
	/**
	* Creates a new ValidationError with structured error messages.
	*
	* @param messages - Structured validation error messages
	* @param options - Optional error options for the base Error class
	*/
	constructor(messages, options) {
		super("Validation failure", options);
		this.messages = messages;
		const ErrorConstructor = this.constructor;
		if ("captureStackTrace" in Error) Error.captureStackTrace(this, ErrorConstructor);
	}
	/**
	* Returns the string tag for this object type.
	*
	* @returns The constructor name of this error
	*/
	get [Symbol.toStringTag]() {
		return this.constructor.name;
	}
	/**
	* Returns a string representation of the validation error.
	*
	* @returns Formatted error string with name, code, and message
	*/
	toString() {
		return `${this.name} [${this.code}]: ${this.message}`;
	}
};
//#endregion
//#region src/errors/main.ts
var main_exports = /* @__PURE__ */ __exportAll({ E_VALIDATION_ERROR: () => E_VALIDATION_ERROR });
/**
* Factory constant for creating ValidationError instances.
* This is the main export for creating validation errors in the Vine validation library.
*
* @example
* throw new E_VALIDATION_ERROR(errors)
*
* @example
* const error = new E_VALIDATION_ERROR({
*   field: 'username',
*   message: 'Username is required'
* })
*/
const E_VALIDATION_ERROR = ValidationError;
//#endregion
//#region src/reporters/simple_error_reporter.ts
/**
* SimpleErrorReporter collects validation error messages as an array of objects.
* It's the default error reporter used by Vine and provides a simple, structured
* format for validation errors.
*
* Each error object contains:
* - message: Human-readable error message
* - field: The field path where validation failed
* - rule: The validation rule that failed
* - index: Array index (for array element errors)
* - meta: Additional error metadata
*
* @example
* const reporter = new SimpleErrorReporter()
* // After validation errors are reported:
* console.log(reporter.errors)
* // [{ message: "Required", field: "email", rule: "required" }]
*/
var SimpleErrorReporter = class {
	/**
	* Flag indicating whether any validation errors have been reported
	*/
	hasErrors = false;
	/**
	* Collection of all reported validation errors
	*/
	errors = [];
	/**
	* Report a validation error by adding it to the errors collection.
	*
	* @param message - The human-readable error message
	* @param rule - The name of the validation rule that failed
	* @param field - The field context containing field path and metadata
	* @param meta - Optional additional metadata about the error
	*/
	report(message, rule, field, meta) {
		const error = {
			message,
			rule,
			field: field.getFieldPath()
		};
		if (meta) error.meta = meta;
		if (field.isArrayMember) error.index = field.name;
		this.hasErrors = true;
		this.errors.push(error);
	}
	/**
	* Creates and returns a ValidationError instance containing all reported errors.
	* This method is called by the validation engine when validation fails.
	*
	* @returns A ValidationError containing all collected error messages
	*/
	createError() {
		return new E_VALIDATION_ERROR(this.errors);
	}
};
//#endregion
export { BOOLEAN_NEGATIVES as a, __exportAll as c, SimpleMessagesProvider as i, main_exports as n, BOOLEAN_POSITIVES as o, ValidationError as r, helpers as s, SimpleErrorReporter as t };
