// @adonisjs/core/build/commands/make/command.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/command.ts
/**
* Command to create a new Ace command class.
* Ace commands are CLI commands that can be executed via the `ace` binary,
* allowing you to create custom functionality for your application's command line interface.
*
* @example
* ```
* ace make:command SendEmails
* ace make:command ProcessPayments
* ace make:command GenerateReports
* ace make:command CleanupFiles
* ```
*/
var MakeCommand = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:command";
	/**
	* The command description
	*/
	static description = "Create a new Ace CLI command class in commands/";
	/**
	* The stub template file to use for generating the command class
	*/
	stubPath = "make/command/main.stub";
	/**
	* Execute the command to create a new Ace command class.
	* Generates the command file with proper CLI command structure.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the command" })], MakeCommand.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeCommand.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeCommand.prototype, "force", void 0);
//#endregion
export { MakeCommand as default };
