// @adonisjs/core/build/commands/make/controller.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
import string from "@poppinss/utils/string";
//#region commands/make/controller.ts
/**
* The make controller command to create an HTTP controller
*
* @example
* ```
* ace make:controller User
* ace make:controller User store update
* ace make:controller User --resource
* ace make:controller User --api
* ace make:controller User --singular
* ```
*/
var MakeController = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:controller";
	/**
	* The command description
	*/
	static description = "Create a new HTTP controller class in app/controllers. Use --resource for CRUD methods or --api for API-only CRUD (no create/edit)";
	/**
	* Command options configuration
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub to use for generating the controller
	*/
	stubPath = "make/controller/main.stub";
	/**
	* Preparing the command state
	*/
	async prepare() {
		/**
		* Use actions stub
		*/
		if (this.actions) this.stubPath = "make/controller/actions.stub";
		/**
		* Use resource stub
		*/
		if (this.resource) if (this.actions) this.logger.warning("Cannot use --resource flag with actions. Ignoring --resource");
		else this.stubPath = "make/controller/resource.stub";
		/**
		* Use api stub
		*/
		if (this.api) if (this.actions) this.logger.warning("Cannot use --api flag with actions. Ignoring --api");
		else this.stubPath = "make/controller/api.stub";
		/**
		* Log warning when both flags are used together
		*/
		if (this.resource && this.api && !this.actions) this.logger.warning("--api and --resource flags cannot be used together. Ignoring --resource");
	}
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			actions: this.actions?.map((action) => string.camelCase(action)),
			entity: this.app.generators.createEntity(this.name),
			singular: this.singular
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "The name of the controller" })], MakeController.prototype, "name", void 0);
__decorate([args.spread({
	description: "Create controller with custom method names",
	required: false
})], MakeController.prototype, "actions", void 0);
__decorate([flags.boolean({
	description: "Generate controller in singular form",
	alias: "s"
})], MakeController.prototype, "singular", void 0);
__decorate([flags.boolean({
	description: "Generate resourceful controller with methods to perform CRUD actions on a resource",
	alias: "r"
})], MakeController.prototype, "resource", void 0);
__decorate([flags.boolean({
	description: "Generate resourceful controller without the \"edit\" and the \"create\" methods",
	alias: "a"
})], MakeController.prototype, "api", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeController.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeController.prototype, "force", void 0);
//#endregion
export { MakeController as default };
