// @adonisjs/core/build/commands/make/event.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/event.ts
/**
* Command to create a new event class.
* Events are data objects that encapsulate information about something
* that happened in your application and can be dispatched to listeners.
*
* @example
* ```
* ace make:event UserRegistered
* ace make:event OrderCompleted
* ace make:event EmailSent
* ```
*/
var MakeEvent = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:event";
	/**
	* The command description
	*/
	static description = "Create a new event class in app/events. Events are dispatched via emitter and handled by listeners";
	/**
	* Command options configuration.
	* Allows unknown flags to be passed through.
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub template file to use for generating the event class
	*/
	stubPath = "make/event/main.stub";
	/**
	* Execute the command to create a new event class.
	* Generates the event file with proper event structure.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the event" })], MakeEvent.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeEvent.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeEvent.prototype, "force", void 0);
//#endregion
export { MakeEvent as default };
