// @adonisjs/core/build/commands/make/exception.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/exception.ts
/**
* Command to create a new custom exception class.
* Custom exceptions allow you to define specific error types for your application
* with custom error messages, status codes, and error handling logic.
*
* @example
* ```
* ace make:exception ValidationException
* ace make:exception UnauthorizedException
* ace make:exception ResourceNotFoundException
* ```
*/
var MakeException = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:exception";
	/**
	* The command description
	*/
	static description = "Create a new custom exception class in app/exceptions with handle and report methods";
	/**
	* Command options configuration.
	* Allows unknown flags to be passed through.
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub template file to use for generating the exception class
	*/
	stubPath = "make/exception/main.stub";
	/**
	* Execute the command to create a new custom exception class.
	* Generates the exception file with proper error handling structure.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the exception" })], MakeException.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeException.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeException.prototype, "force", void 0);
//#endregion
export { MakeException as default };
