// @adonisjs/core/build/commands/make/listener.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/listener.ts
/**
* Command to create a new event listener class.
* Event listeners handle events dispatched by the application and can optionally
* generate the corresponding event class automatically.
*
* @example
* ```
* ace make:listener UserRegistered
* ace make:listener EmailSent --event=EmailSent
* ace make:listener OrderCompleted --event=OrderEvent
* ```
*/
var MakeListener = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:listener";
	/**
	* The command description
	*/
	static description = "Create a new event listener class in app/listeners. Use --event to also generate the event class and bind them together";
	/**
	* Command options configuration.
	* Allows unknown flags to be passed through.
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub template file to use for generating the event listener
	*/
	stubPath = "make/listener/main.stub";
	/**
	* Prepare the command by selecting the appropriate stub based on options.
	* Uses a different stub when generating a listener for a specific event.
	*/
	prepare() {
		if (this.event) this.stubPath = "make/listener/for_event.stub";
	}
	/**
	* Execute the command to create a new event listener.
	* If an event is specified, creates the event class first,
	* then generates the listener with proper event binding.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		if (this.event) {
			const { exitCode } = await this.kernel.exec("make:event", [this.event]);
			/**
			* Create listener only when make:event is completed successfully
			*/
			if (exitCode === 0) {
				const eventEntity = this.app.generators.createEntity(this.event);
				await codemods.makeUsingStub(stubsRoot, this.stubPath, {
					event: eventEntity,
					flags: this.parsed.flags,
					entity: this.app.generators.createEntity(this.name)
				}, { contentsFromFile: this.contentsFrom });
			}
			return;
		}
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the event listener" })], MakeListener.prototype, "name", void 0);
__decorate([flags.string({
	description: "Generate an event class alongside the listener",
	alias: "e"
})], MakeListener.prototype, "event", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeListener.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeListener.prototype, "force", void 0);
//#endregion
export { MakeListener as default };
