// @adonisjs/core/build/commands/make/middleware.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { t as stringHelpers } from "../../string-8_8He-HH.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
import string from "@poppinss/utils/string";
import { basename, extname, relative } from "node:path";
//#region commands/make/middleware.ts
/**
* The make middleware command to create a new middleware
* class.
*
* @example
* ```
* ace make:middleware Auth
* ace make:middleware Auth --stack=server
* ace make:middleware Auth --stack=named
* ace make:middleware Auth --stack=router
* ```
*/
var MakeMiddleware = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:middleware";
	/**
	* The command description
	*/
	static description = "Create a new middleware class in app/middleware and register it in start/kernel.ts under the chosen stack (server, router, or named)";
	/**
	* Command options configuration
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub to use for generating the middleware
	*/
	stubPath = "make/middleware/main.stub";
	async run() {
		const stackChoices = [
			"server",
			"router",
			"named"
		];
		/**
		* Prompt to select the stack under which to register
		* the middleware
		*/
		if (!this.stack) this.stack = await this.prompt.choice("Under which stack you want to register the middleware?", stackChoices);
		/**
		* Error out when mentioned stack is invalid
		*/
		if (!stackChoices.includes(this.stack)) {
			this.exitCode = 1;
			this.logger.error(`Invalid middleware stack "${this.stack}". Select from "${stackChoices.join(", ")}"`);
			return;
		}
		/**
		* Create middleware
		*/
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
		/**
		* Creative relative path for the middleware file from
		* the "./app/middleware" directory
		*/
		const middlewareRelativePath = stringHelpers.toUnixSlash(relative(this.app.middlewarePath(), destination).replace(extname(destination), ""));
		/**
		* Take the middleware relative path, remove `_middleware` prefix from it
		* and convert everything to camelcase
		*/
		const name = string.camelCase(basename(middlewareRelativePath).replace(/_middleware$/, ""));
		/**
		* Register middleware
		*/
		await codemods.registerMiddleware(this.stack, [{
			name,
			path: `#middleware/${middlewareRelativePath}`
		}]);
	}
};
__decorate([args.string({ description: "Name of the middleware" })], MakeMiddleware.prototype, "name", void 0);
__decorate([flags.string({
	description: "The stack in which to register the middleware",
	alias: "s"
})], MakeMiddleware.prototype, "stack", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeMiddleware.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeMiddleware.prototype, "force", void 0);
//#endregion
export { MakeMiddleware as default };
