// @adonisjs/core/build/commands/make/preload.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { t as stringHelpers } from "../../string-8_8He-HH.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
import { extname, relative } from "node:path";
//#region commands/make/preload.ts
const ALLOWED_ENVIRONMENTS = [
	"web",
	"console",
	"test",
	"repl"
];
/**
* Command to create a new preload file in the start directory.
* Preload files are executed during application startup and can be used
* to set up global configurations, register global bindings, or perform
* application-wide initialization tasks.
*
* @example
* ```
* ace make:preload routes
* ace make:preload database --register
* ace make:preload events --no-register
* ace make:preload kernel --environments=web,console
* ```
*/
var MakePreload = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:preload";
	/**
	* The command description
	*/
	static description = "Create a new preload file in start/ and optionally register it in adonisrc.ts. Preload files run during app boot";
	/**
	* The stub template file to use for generating the preload file
	*/
	stubPath = "make/preload/main.stub";
	/**
	* Validate that all specified environments are valid application environments.
	*
	* @returns True if all environments are valid or none specified, false otherwise
	*/
	#isEnvironmentsFlagValid() {
		if (!this.environments || !this.environments.length) return true;
		return this.environments.every((one) => ALLOWED_ENVIRONMENTS.includes(one));
	}
	/**
	* Execute the command to create a new preload file.
	* Validates inputs, generates the preload file, and optionally registers it in .adonisrc.ts.
	*/
	async run() {
		/**
		* Ensure the environments are valid when provided via flag
		*/
		if (!this.#isEnvironmentsFlagValid()) {
			this.logger.error(`Invalid environment(s) "${this.environments}". Only "${ALLOWED_ENVIRONMENTS}" are allowed`);
			return;
		}
		/**
		* Display prompt to know if we should register the preload
		* file inside the ".adonisrc.ts" file.
		*/
		if (this.register === void 0) this.register = await this.prompt.confirm("Do you want to register the preload file in .adonisrc.ts file?");
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
		/**
		* Do not register when prompt has been denied or "--no-register"
		* flag was used
		*/
		if (!this.register) return;
		/**
		* Creative relative path for the preload file from
		* the "./start" directory
		*/
		const preloadFileRelativePath = stringHelpers.toUnixSlash(relative(this.app.startPath(), destination).replace(extname(destination), ""));
		await codemods.updateRcFile((rcFile) => {
			rcFile.addPreloadFile(`#start/${preloadFileRelativePath}`, this.environments);
		});
	}
};
__decorate([args.string({ description: "Name of the preload file" })], MakePreload.prototype, "name", void 0);
__decorate([flags.boolean({
	description: "Auto register the preload file inside the .adonisrc.ts file",
	showNegatedVariantInHelp: true,
	alias: "r"
})], MakePreload.prototype, "register", void 0);
__decorate([flags.array({
	description: `Define the preload file's environment. Accepted values are "${ALLOWED_ENVIRONMENTS}"`,
	alias: "e"
})], MakePreload.prototype, "environments", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakePreload.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakePreload.prototype, "force", void 0);
//#endregion
export { MakePreload as default };
