// @adonisjs/core/build/commands/make/provider.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { t as stringHelpers } from "../../string-8_8He-HH.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
import { extname, relative } from "node:path";
//#region commands/make/provider.ts
const ALLOWED_ENVIRONMENTS = [
	"web",
	"console",
	"test",
	"repl"
];
/**
* Command to create a new service provider class.
* Service providers are used to register bindings, configure services,
* and bootstrap application components during startup.
*
* @example
* ```
* ace make:provider AuthProvider
* ace make:provider DatabaseProvider --register
* ace make:provider AppProvider --no-register
* ace make:provider CacheProvider --environments=web,console
* ```
*/
var MakeProvider = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:provider";
	/**
	* The command description
	*/
	static description = "Create a new service provider in providers/ and optionally register it in adonisrc.ts. Providers register IoC container bindings";
	/**
	* The stub template file to use for generating the provider class
	*/
	stubPath = "make/provider/main.stub";
	/**
	* Validate that all specified environments are valid application environments.
	*
	* @returns True if all environments are valid or none specified, false otherwise
	*/
	#isEnvironmentsFlagValid() {
		if (!this.environments || !this.environments.length) return true;
		return this.environments.every((one) => ALLOWED_ENVIRONMENTS.includes(one));
	}
	/**
	* Execute the command to create a new service provider.
	* Validates inputs, generates the provider file, and optionally registers it in .adonisrc.ts.
	*/
	async run() {
		/**
		* Ensure the environments are valid when provided via flag
		*/
		if (!this.#isEnvironmentsFlagValid()) {
			this.logger.error(`Invalid environment(s) "${this.environments}". Only "${ALLOWED_ENVIRONMENTS}" are allowed`);
			return;
		}
		/**
		* Display prompt to know if we should register the provider
		* file inside the ".adonisrc.ts" file.
		*/
		if (this.register === void 0) this.register = await this.prompt.confirm("Do you want to register the provider in .adonisrc.ts file?");
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
		/**
		* Do not register when prompt has been denied or "--no-register"
		* flag was used
		*/
		if (!this.register) return;
		/**
		* Creative relative path for the provider file from
		* the "./start" directory
		*/
		const providerRelativePath = stringHelpers.toUnixSlash(relative(this.app.providersPath(), destination).replace(extname(destination), ""));
		await codemods.updateRcFile((rcFile) => {
			rcFile.addProvider(`#providers/${providerRelativePath}`, this.environments);
		});
	}
};
__decorate([args.string({ description: "Name of the provider" })], MakeProvider.prototype, "name", void 0);
__decorate([flags.boolean({
	description: "Auto register the provider inside the .adonisrc.ts file",
	showNegatedVariantInHelp: true,
	alias: "r"
})], MakeProvider.prototype, "register", void 0);
__decorate([flags.array({
	description: `Define the provider environment. Accepted values are "${ALLOWED_ENVIRONMENTS}"`,
	alias: "e"
})], MakeProvider.prototype, "environments", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeProvider.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeProvider.prototype, "force", void 0);
//#endregion
export { MakeProvider as default };
