// @adonisjs/core/build/commands/make/service.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/service.ts
/**
* Make a new service class
*
* @example
* ```
* ace make:service UserService
* ace make:service AuthService
* ace make:service User/ProfileService
* ```
*/
var MakeService = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:service";
	/**
	* The command description
	*/
	static description = "Create a new service class in app/services. Services encapsulate reusable business logic outside of controllers";
	/**
	* Command options configuration
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub to use for generating the service class
	*/
	stubPath = "make/service/main.stub";
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the service" })], MakeService.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeService.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeService.prototype, "force", void 0);
//#endregion
export { MakeService as default };
