// @adonisjs/core/build/commands/make/test.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/test.ts
/**
* Command to create a new Japa test file.
* Supports multiple test suites and automatically detects or prompts for
* the appropriate suite and directory based on application configuration.
*
* @example
* ```
* ace make:test UserController
* ace make:test UserModel --suite=unit
* ace make:test AuthService --suite=integration
* ```
*/
var MakeTest = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:test";
	/**
	* The command description
	*/
	static description = "Create a new Japa test file in tests/. Use --suite to specify the test suite (unit, functional, etc.)";
	/**
	* The stub template file to use for generating the test file
	*/
	stubPath = "make/test/main.stub";
	/**
	* Determine the test suite name for creating the test file.
	* Uses the provided suite flag, or automatically selects if only one suite exists,
	* or prompts the user to choose from available suites.
	*
	* @returns The name of the selected test suite
	*/
	async #getSuite() {
		if (this.suite) return this.suite;
		/**
		* Use the first suite from the rcFile when there is only
		* one suite
		*/
		const rcFileSuites = this.app.rcFile.tests.suites;
		if (rcFileSuites.length === 1) return rcFileSuites[0].name;
		/**
		* Prompt the user to select a suite manually
		*/
		return this.prompt.choice("Select the suite for the test file", this.app.rcFile.tests.suites.map((suite) => {
			return suite.name;
		}), { validate(choice) {
			return choice ? true : "Please select a suite";
		} });
	}
	/**
	* Determine the directory path for the test file within the selected suite.
	* Automatically selects if only one directory exists, otherwise prompts the user.
	*
	* @param directories - Array of available directories for the suite
	* @returns The selected directory path
	*/
	async #getSuiteDirectory(directories) {
		if (directories.length === 1) return directories[0];
		return this.prompt.choice("Select directory for the test file", directories, { validate(choice) {
			return choice ? true : "Please select a directory";
		} });
	}
	/**
	* Find suite configuration from the RC file by name.
	*
	* @param suiteName - The name of the suite to find
	* @returns The suite configuration or undefined if not found
	*/
	#findSuite(suiteName) {
		return this.app.rcFile.tests.suites.find((suite) => {
			return suite.name === suiteName;
		});
	}
	/**
	* Execute the command to create a new test file.
	* Validates the suite exists, prompts for missing information,
	* and generates the test file in the appropriate location.
	*/
	async run() {
		const suite = this.#findSuite(await this.#getSuite());
		/**
		* Show error when mentioned/selected suite does not exist
		*/
		if (!suite) {
			this.logger.error(`The "${this.suite}" suite is not configured inside the "adonisrc.js" file`);
			this.exitCode = 1;
			return;
		}
		/**
		* Generate entity
		*/
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name),
			suite: { directory: await this.#getSuiteDirectory(suite.directories) }
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the test file" })], MakeTest.prototype, "name", void 0);
__decorate([flags.string({
	description: "The suite for which to create the test file",
	alias: "s"
})], MakeTest.prototype, "suite", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeTest.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeTest.prototype, "force", void 0);
//#endregion
export { MakeTest as default };
