// @adonisjs/core/build/commands/make/transformer.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/transformer.ts
/**
* Command to create a new transformer class.
* Transformers are used to serialize data objects (like models) into specific
* formats for API responses, allowing you to control which fields are exposed
* and how data is structured for clients.
*
* @example
* ```
* ace make:transformer User
* ace make:transformer Post
* ace make:transformer ProductTransformer
* ```
*/
var MakeTransformer = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:transformer";
	/**
	* The command description
	*/
	static description = "Create a new transformer class in app/transformers for serializing data in API responses";
	/**
	* Command options configuration.
	* Allows unknown flags to be passed through.
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub template file to use for generating the transformer class
	*/
	stubPath = "make/transformer/main.stub";
	/**
	* Execute the command to create a new transformer class.
	* Generates the transformer file with proper data serialization structure.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name),
			model: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Entity name for which to generate the transformer" })], MakeTransformer.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeTransformer.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeTransformer.prototype, "force", void 0);
//#endregion
export { MakeTransformer as default };
