// @adonisjs/core/build/commands/make/validator.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/validator.ts
/**
* Command to create a new VineJS validator file.
* Validators define reusable validation schemas for request validation,
* and can be generated as simple validators or resource-based validators
* with create and update schemas.
*
* @example
* ```
* ace make:validator UserValidator
* ace make:validator PostValidator --resource
* ace make:validator ContactValidator
* ```
*/
var MakeValidator = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:validator";
	/**
	* The command description
	*/
	static description = "Create a new VineJS validator file in app/validators. Use --resource to generate both create and update validation schemas";
	/**
	* Command options configuration.
	* Allows unknown flags to be passed through.
	*/
	static options = { allowUnknownFlags: true };
	/**
	* The stub template file to use for generating the validator
	*/
	stubPath = "make/validator/main.stub";
	/**
	* Prepare the command by selecting the appropriate stub based on options.
	* Uses a resource stub when generating validators for CRUD operations.
	*/
	async prepare() {
		/**
		* Use resource stub
		*/
		if (this.resource) this.stubPath = "make/validator/resource.stub";
	}
	/**
	* Execute the command to create a new VineJS validator file.
	* Generates the validator with the appropriate stub template.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the validator file" })], MakeValidator.prototype, "name", void 0);
__decorate([flags.boolean({ description: "Create a file with pre-defined validators for create and update actions" })], MakeValidator.prototype, "resource", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeValidator.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeValidator.prototype, "force", void 0);
//#endregion
export { MakeValidator as default };
