// @adonisjs/core/build/commands/make/view.js
import { t as stubsRoot } from "../../main-CfTiGsiS.js";
import { o as args, p as BaseCommand, u as flags } from "../../main-DzNXtXg3.js";
import { t as __decorate } from "../../decorate-yKe-VEeE.js";
//#region commands/make/view.ts
/**
* Command to create a new Edge.js template file.
* Edge templates are used for rendering HTML views in your web application,
* supporting layouts, partials, components, and template inheritance.
*
* @example
* ```
* ace make:view home
* ace make:view users/profile
* ace make:view components/navbar
* ace make:view layouts/app
* ```
*/
var MakeView = class extends BaseCommand {
	/**
	* The command name
	*/
	static commandName = "make:view";
	/**
	* The command description
	*/
	static description = "Create a new Edge.js template file in resources/views";
	/**
	* The stub template file to use for generating the Edge template
	*/
	stubPath = "make/view/main.stub";
	/**
	* Execute the command to create a new Edge.js template file.
	* Generates the template file in the views directory.
	*/
	async run() {
		const codemods = await this.createCodemods();
		codemods.overwriteExisting = this.force === true;
		await codemods.makeUsingStub(stubsRoot, this.stubPath, {
			flags: this.parsed.flags,
			entity: this.app.generators.createEntity(this.name)
		}, { contentsFromFile: this.contentsFrom });
	}
};
__decorate([args.string({ description: "Name of the template" })], MakeView.prototype, "name", void 0);
__decorate([flags.string({ description: "Use the contents of the given file as the generated output" })], MakeView.prototype, "contentsFrom", void 0);
__decorate([flags.boolean({
	description: "Forcefully overwrite existing files",
	alias: "f"
})], MakeView.prototype, "force", void 0);
//#endregion
export { MakeView as default };
