// @adonisjs/core/build/modules/encryption/drivers/legacy.js
import { t as E_BLIND_INDEX_NOT_SUPPORTED } from "../../../errors-BScZTzBu.js";
import { BaseDriver, Hmac, base64UrlDecode, base64UrlEncode, errors } from "@boringnode/encryption";
import { MessageBuilder } from "@poppinss/utils";
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
//#region modules/encryption/drivers/legacy.ts
/**
* Factory function to create a Legacy encryption configuration.
*
* @example
* ```ts
* drivers.legacy({
*   keys: [env.get('APP_KEY')]
* })
* ```
*/
function legacy(config) {
	return {
		driver: (key) => new Legacy({ key }),
		keys: config.keys
	};
}
/**
* Legacy encryption driver for AdonisJS.
*
* This driver maintains compatibility with the old AdonisJS v6 encryption
* format. It uses:
* - AES-256-CBC for encryption
* - HMAC SHA-256 for integrity verification
* - MessageBuilder from @poppinss/utils for encoding values
*
* Encrypted format: `[encrypted_base64url].[iv_base64url].[hmac]`
*
* @example
* ```ts
* const driver = new Legacy({ key: 'your-32-character-secret-key!!' })
*
* const encrypted = driver.encrypt('sensitive data')
* const decrypted = driver.decrypt(encrypted)
* ```
*/
var Legacy = class extends BaseDriver {
	constructor(config) {
		super(config);
		/**
		* The key length must be at least 16 characters
		*/
		if (this.cryptoKey.length < 16) throw new errors.E_INSECURE_ENCRYPTER_KEY();
	}
	encrypt(payload, expiresInOrOptions, purpose) {
		let expiresIn;
		let actualPurpose;
		if (typeof expiresInOrOptions === "object" && expiresInOrOptions !== null) {
			expiresIn = expiresInOrOptions.expiresIn;
			actualPurpose = expiresInOrOptions.purpose;
		} else {
			expiresIn = expiresInOrOptions;
			actualPurpose = purpose;
		}
		const iv = randomBytes(16);
		const cipher = createCipheriv("aes-256-cbc", this.cryptoKey.subarray(0, 32), iv);
		const plainText = new MessageBuilder().build(payload, expiresIn, actualPurpose);
		const macPayload = `${base64UrlEncode(Buffer.concat([cipher.update(plainText), cipher.final()]))}${this.separator}${base64UrlEncode(iv)}`;
		const hmac = new Hmac(this.cryptoKey).generate(macPayload);
		return this.computeReturns([macPayload, hmac]);
	}
	/**
	* Decrypt value and verify it against a purpose
	*/
	decrypt(value, purpose) {
		if (typeof value !== "string") return null;
		const [cipherEncoded, ivEncoded, macEncoded] = value.split(this.separator);
		if (!cipherEncoded || !ivEncoded || !macEncoded) return null;
		const cipherText = base64UrlDecode(cipherEncoded);
		if (!cipherText) return null;
		const iv = base64UrlDecode(ivEncoded);
		if (!iv) return null;
		if (!new Hmac(this.cryptoKey).compare(`${cipherEncoded}${this.separator}${ivEncoded}`, macEncoded)) return null;
		try {
			const decipher = createDecipheriv("aes-256-cbc", this.cryptoKey.subarray(0, 32), iv);
			const plainTextBuffer = Buffer.concat([decipher.update(cipherText), decipher.final()]);
			return new MessageBuilder().verify(plainTextBuffer, purpose);
		} catch {
			return null;
		}
	}
	/**
	* Legacy driver does not support blind indexes.
	*/
	blindIndex(_payload, _purpose) {
		throw new E_BLIND_INDEX_NOT_SUPPORTED(["legacy"]);
	}
	/**
	* Legacy driver does not support blind indexes.
	*/
	blindIndexes(_payload, _purpose) {
		throw new E_BLIND_INDEX_NOT_SUPPORTED(["legacy"]);
	}
};
//#endregion
export { Legacy, legacy };
