// @adonisjs/core/build/src/helpers/types.js
import is from "@sindresorhus/is";
//#region src/helpers/types.ts
/**
* @deprecated
* Use "is" helpers instead. The types helpers exist for backward compatibility.
*
* @example
* // Instead of using types helpers (deprecated)
* types.isString('hello')
* types.isNumber(42)
*
* @example
* // Use the new "is" helpers
* import { is } from '@adonisjs/core/helpers'
* is.string('hello')
* is.number(42)
*/
const types = {
	lookup: is,
	isNull: is.null,
	isBoolean: is.boolean,
	isBuffer: is.buffer,
	isNumber: is.number,
	isString: is.string,
	isArguments: is.arguments,
	isObject: is.object,
	isDate: is.date,
	isArray: is.array,
	isRegexp: is.regExp,
	isError: is.error,
	isFunction: is.function,
	isClass: is.class,
	isInteger: is.integer,
	isFloat(value) {
		return value !== (value | 0);
	},
	isDecimal(value, options) {
		if (typeof value === "number") value = value.toString();
		const decimalPlaces = options && options.decimalPlaces || "1,";
		return new RegExp(`^[-+]?([0-9]+)?(\\.[0-9]{${decimalPlaces}})$`).test(value);
	}
};
//#endregion
export { types as default };
